spec/controllers/projects/classifications_controller_spec.rb

```diff
@@ -0,0 +1,208 @@
+require 'spec_helper'
+
+spec_meta = {
+  type: :controller,
+  feature: :projects
+}
+
+RSpec.describe Projects::ClassificationsController, spec_meta do
+  shared_context 'project can be found' do
+    let(:project) { FactoryBot.create(:project) }
+
+    before do
+      allow(Project).to receive(:find).with('1').and_return(project)
+    end
+  end
+
+  shared_context 'request can be found' do
+    include_context 'project can be found'
+
+    let(:info_request) { FactoryBot.create(:info_request, user: user) }
+
+    before do
+      info_requests = double(:info_requests_collection)
+      allow(project).to receive(:info_requests).and_return(info_requests)
+      allow(info_requests).to receive(:find_by!).with(url_title: 'url_title').
+        and_return(info_request)
+    end
```

Comment:
The problem with not stubbing like:
```ruby
allow(info_requests).to receive(:find_by!).with(url_title: 'url_title').and_return(info_request)
```
is that we can't then do:
```ruby
expect(info_request).to receive(:set_described_state)
```
this leads us to needing to knowledge about and/or testing the implementation of `InfoRequest#set_described_state` in the wrong place.

Reply:
hmm yeah not ideal.

Would be nice if you could use a dynamic finder for like `@project.find_info_request_by_url_title(params[:url_title])` – at least then you're not having to stub an entire method chain.

Happy to leave as-is for now, but its something I'd like to find a better solution to in general.

Reply:
Not suggesting this for now, but might be able to actually build this ourselves:

```ruby
  def method_missing(method_id, *arguments)
    if match = /find_(?<association>[_a-zA-Z]\w+)_by_(?<attribute>[_a-zA-Z]\w+)/.match(method_id.to_s)
      return super unless respond_to?(match[:association].pluralize)
      send(match[:association].pluralize).send("find_by_#{ match[:attribute] }", *arguments)
    else
      super
    end
  end
```

```ruby
project
# => #<Project:0x00005576e62ef768 …>
project.info_requests.first.url_title
# => "skjdhbc_ohlcc"
project.find_info_request_by_url_title("skjdhbc_ohlcc").id
#  InfoRequest Load (4.0ms)  SELECT  "info_requests".* FROM "info_requests" LEFT JOIN project_resources r1 ON r1.resource_id = info_requests.id AND r1.resource_type = 'InfoRequest' LEFT JOIN project_resources r2 ON r2.resource_id = info_requests.info_request_batch_id AND r2.resource_type = 'InfoRequestBatch' WHERE (r1.project_id = 1 OR r2.project_id = 1) AND "info_requests"."url_title" = $1 LIMIT $2  [["url_title", "skjdhbc_ohlcc"], ["LIMIT", 1]]
# => 113

```

Reply:
In fact, this can be simplified way more to just be:

```ruby
# model
def method_missing(method_id, *arguments)
  if match = /find_(?<association>[_a-zA-Z]\w+)_by/.match(method_id.to_s)
    return super unless respond_to?(match[:association].pluralize)
    send(match[:association].pluralize).send(:find_by, *arguments)
  else
    super
  end
end

# controller
@project.find_info_request_by(url_title: params[:url_title])

# controller spec
before do
  allow(project).
    to receive(:find_info_request_by).
      with(url_title: info_request.url_title).
        and_return(info_request)
end
```

Just posting here for future reference